pmd-core/src/main/java/net/sourceforge/pmd/util/document/DocumentOperationsApplierForNonOverlappingRegions.java

```diff
@@ -0,0 +1,102 @@
+
+public class DocumentOperationsApplierForNonOverlappingRegions {
+
+    private final Document document;
+    private final List<DocumentOperation> operations;
+    private final Comparator<DocumentOperation> comparator;
+
+    private boolean applied;
+
+    public DocumentOperationsApplierForNonOverlappingRegions(final Document document) {
+        this.document = Objects.requireNonNull(document);
+        comparator = new DocumentOperationNonOverlappingRegionsComparator();
+        operations = new ArrayList<>();
+        applied = false;
+    }
+
+    public void addDocumentOperation(DocumentOperation documentOperation) {
+        assertOperationsHaveNotBeenApplied();
+
+        final int index = getIndexForDocumentOperation(Objects.requireNonNull(documentOperation));
+        operations.add(index, documentOperation);
+    }
+
+    private void assertOperationsHaveNotBeenApplied() {
+        if (applied) {
+            throw new IllegalStateException("Document operations have already been applied to the document");
+        }
+    }
+
+    private int getIndexForDocumentOperation(final DocumentOperation documentOperation) {
+        int potentialIndex = Collections.binarySearch(operations, documentOperation, comparator);
+
+        if (potentialIndex < 0) {
+            return ~potentialIndex;
+        }
+
+        final int lastIndex = operations.size() - 1;
+        while (potentialIndex < lastIndex && areSiblingsEqual(potentialIndex)) {
+            potentialIndex++;
+        }
+        return potentialIndex + 1;
+    }
+
+    private boolean areSiblingsEqual(final int index) {
+        return comparator.compare(operations.get(index), operations.get(index + 1)) == 0;
+    }
+
+    public void apply() {
+        assertOperationsHaveNotBeenApplied();
+        applied = true;
+
+        for (final DocumentOperation operation : operations) {
+            operation.apply(document);
+        }
+    }
+
+    private class DocumentOperationNonOverlappingRegionsComparator implements Comparator<DocumentOperation> {
+
+        @Override
+        public int compare(final DocumentOperation o1, final DocumentOperation o2) {
+            final RegionByLine r1 = Objects.requireNonNull(o1).getRegionByLine();
+            final RegionByLine r2 = Objects.requireNonNull(o2).getRegionByLine();
+
+            final int comparison;
+            if (areInsertOperations(r1, r2)) {
+                comparison = 0;
+            } else if (doesFirstRegionEndBeforeSecondRegionBegins(r1, r2)) {
+                comparison = -1;
+            } else if (doesFirstRegionEndBeforeSecondRegionBegins(r2, r1)) {
+                comparison = 1;
+            } else {
+                throw new IllegalArgumentException("Regions between document operations overlap, " + r1.toString() + "\n" + r2.toString());
+            }
+            return comparison;
+        }
+
+        private boolean areInsertOperations(final RegionByLine r1, final RegionByLine r2) {
```

Comment:
`areInsertOperations` doesn't check if the operation is an insert?

Reply:
The name of the method is misleading, this method really checks for two insert operations at the same offset in the file.

Reply:
nothing guarantees such assumption

Reply:
An insert operation doesn't take a region of length greater than zero in the file, but it has an offset to where insert it. For both operations to be an insertion at the same offset they have to start at the same line and column and their length representation at the file to be zero (i.e. beginLine == endLine and beginColumn == endColumn). This is what the method is checking.

Reply:
That's an implementation detail of an insert operation that:
1. this class shouldn't know
2. this class can't guarantee will not be true for any other type of operation

Reply:
1. By the class you mean the DocumentOperationNonOverlappingRegionsComparator or DocumentOperationsApplierForNonOverlappingRegions?
2. It is true that you can create a replace/delete operation with the same region as an insert operation, although it would not make any sense in a real case scenario. The underlying question in this method would be: if both start at the same offset, will they bother each other? If both operations don't start at the same offset, then I have to check that one starts after the other one finishes. If both operations start at the same offset and have length zero then they can be put one after the other, because they won't overlap.

Reply:
1. For this scenario, neither.
2. If that's what your method actually checks, that's how this method should be called.

Reply:
As a solution I can delegate the comparison of operations to the comparison of their regions by using the comparable interface implemented in a region.